### PR TITLE
No result in read specification & make it really immutable

### PR DESCRIPTION
**Pull request by mpraglowski**
Fix for #398 and #407

### CODE REVIEW

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 1]
```diff
@@ -22,15 +22,15 @@ def stream_name
...
+      start.equal?(:head)
     end
 
     def forward?
       @direction.equal?(:forward)
     end
 
     def backward?
-      !forward?
+      @direction.equal?(:backward)
```
Comment (pawelpacana) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: `start` over `@start`, `@direction` over `direction`?

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 2]
```diff
@@ -4,68 +4,82 @@ module RubyEventStore
...
-    # @private
-    attr_reader :result
+    def last?
+      @read_as.equal?(:last)
+    end
 
-    # @api private
-    # @private
-    def initialize(repository, mapper, result = Result.new(:forward, :head, nil, Stream.new(GLOBAL_STREAM), :all, DEFAULT_BATCH_SIZE))
+    def initialize(repository, mapper,
```
Comment (paneq) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: Why did you remove the `private` marker? We don't want to document this method, it is not intended to be used publicly by RES users.

Hmmm. It seems this we dropped `Result` class and now this is a `Specification` initializer. Would be nice to move it to the top.

Comment (paneq) [reply ruby_event_store/lib/ruby_event_store/specification.rb]: I would consider if `Specification` is a private class, not intended for public documentation.

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 3]
```diff
@@ -4,68 +4,82 @@ module RubyEventStore
...
     # Limits the query to certain stream.
     # {http://railseventstore.org/docs/read/ Find out more}.
     #
-    # @param stream_name [String] name of the stream to get events from
+    # @param name [String] name of the stream to get events from
     # @return [Specification]
-    def stream(stream_name)
-      Specification.new(repository, mapper, result.dup.tap { |r| r.stream = Stream.new(stream_name) })
+    def stream(name)
+      Specification.new(repository, mapper,
```
Comment (paneq) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: quite verbose. I would consider something along the line of:

```
new = self.dup
new.instance_variable_set(:@stream, name)
new.freeze
new
```

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 4]
```diff
@@ -190,10 +245,49 @@ def first
...
+    # determine if two objects reference the same hash key.
+    #
+    # This hash is based on
+    # * class
+    # * stream_name
+    # * start
+    # * direction
+    # * read_as
+    # * batch_size
+    def hash
```
Comment (paneq) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: Could we use all `instance_variables` ? Would that be safer? It's hard for me to say if we forgot to include something or not.

**Comment by mpraglowski** [comment]
This was a step in wrong direction. Let's try smaller steps like https://github.com/RailsEventStore/rails_event_store/pull/415
